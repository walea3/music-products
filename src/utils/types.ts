export type Product = {
  id: number;
  title: string;
  description: string;
  artist: string;
  isFree: boolean;
  image1: string;
  image2: string;
  accentColor: string;
  backgroundColor: string;
  textColor: string;
  feature_order: string;
  price?: string;
};
