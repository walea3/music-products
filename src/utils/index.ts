export * from './product';
export * from './useAPI';
export * from './test';
