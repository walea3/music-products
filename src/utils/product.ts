export type Product = {
  id: number;
  title: string;
  description: string;
  artist: string;
  isFree: boolean;
  image1: string;
  image2: string;
  accentColor: string;
  backgroundColor: string;
  textColor: string;
  feature_order: string;
  price?: string;
};

export const BASE_PRICE = 12.99;
export const DEFAULT_BG_COLOR = '#424242';

export let naiveProductCache: { [id: number]: Product } = {};

export const hasMismatchedData = ({ id }: Product) => mismatchedDataMap[id];

export const parseProducts = (products: Product[]) => {
  naiveProductCache = products.reduce(
    (cache, product) =>
      hasMismatchedData(product) ? { ...cache, [product.id]: product } : cache,
    naiveProductCache,
  );

  return Object.keys(naiveProductCache).length
    ? products.map(fixMismatch)
    : products;
};

export const fixMismatch = (product: Product) => {
  const mismatches = mismatchedDataMap[product.id];

  return mismatches
    ? mismatches.reduce((fixedProduct, [mismatchId, ...fields]) => {
        const matchedProduct = naiveProductCache[mismatchId];
        return fields.reduce(
          (acc, field) => ({ ...acc, [field]: matchedProduct[field] }),
          fixedProduct,
        );
      }, product)
    : product;
};

const mismatchedDataMap: {
  [id: number]: [mismatchId: number, ...fields: (keyof Product)[]][];
} = {
  111: [[114, 'image1', 'image2']],
  114: [[111, 'image1', 'image2']],
};
