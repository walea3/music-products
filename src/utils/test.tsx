import { ReactElement } from 'react';
import { render } from '@testing-library/react';
import { Route, RouteProps, BrowserRouter } from 'react-router-dom';

export const renderWithRouter = (ui: ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return render(ui, { wrapper: BrowserRouter });
};

export const renderPage = (
  page: RouteProps['children'],
  path: RouteProps['path'],
  route: string,
) => renderWithRouter(<Route path={path}>{page}</Route>, { route });

// based on react-testing library default
const TEST_ID_ATTR = 'data-testid';
const TEST_ID_DELIMITER = '-';

// @TODO if desired remove test id in prod
export const testIdProp = (baseTestId: string) => (id?: string) => ({
  [TEST_ID_ATTR]: `${baseTestId}${id ? `${TEST_ID_DELIMITER}${id}` : ''}`,
});

export const getTestId = (...args: any[]) =>
  args.filter(Boolean).join(TEST_ID_DELIMITER);
