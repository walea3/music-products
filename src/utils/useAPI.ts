import { useState, useEffect, useCallback } from 'react';

export const useAPI = <T>(url: string) => {
  const [error, setError] = useState(null);
  const [data, setData] = useState<T>();
  const [isLoading, setLoading] = useState(true);

  const fetchData = useCallback(async () => {
    setLoading(true);
    setError(null);

    try {
      const res = await fetch(`/api/${url}`);
      const data: T = await res.json();

      setData(data);
    } catch (error) {
      setError(error);
    } finally {
      setLoading(false);
    }
  }, [url]);

  useEffect(() => {
    fetchData();
  }, [fetchData]);

  return [isLoading, data, error, fetchData] as const;
};
