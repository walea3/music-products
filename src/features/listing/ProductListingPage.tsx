import { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import orderBy from 'lodash/orderBy';

import { Product, useAPI, testIdProp } from 'utils';

import { ProductItem } from './components/ProductItem';

export const getProductUrl = ({ id }: Product) => `/product-${id}`;

export function ProductListingPage() {
  const [products, setProducts] = useState<Product[]>([]);
  const [isLoading, data = [], error, fetchProducts] = useAPI<Product[]>(
    'products',
  );

  useEffect(() => {
    if (data.length > 0) setProducts(orderBy(data, ['feature_order', 'asc']));
  }, [data]);

  const testId = testIdProp('Products');

  return (
    <div>
      <h1>Products</h1>

      {error && (
        <>
          There was problem getting you products.{' '}
          <button onClick={fetchProducts} {...testId('tryAgain')}>
            Try again.
          </button>
        </>
      )}

      {isLoading && <div>Loading products ...</div>}

      {products.length < 1 ? (
        <div>No products, yet.</div>
      ) : (
        <ProductList>
          {products.map((product) => (
            <ProductListItem key={product.id}>
              <Link to={getProductUrl(product)}>
                <ProductItem {...product} />
              </Link>
            </ProductListItem>
          ))}
        </ProductList>
      )}
    </div>
  );
}

const ProductList = styled.ul`
  text-align: center;
  list-style: none;
  margin: 0 auto;
`;

const ProductListItem = styled.li`
  display: inline-block;
  margin: 10px;
  vertical-align: top;
`;
