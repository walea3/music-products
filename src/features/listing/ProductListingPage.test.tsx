import React from 'react';
import { fireEvent, RenderResult } from '@testing-library/react';

import { getTestId, useAPI, renderWithRouter } from 'utils';

import { ProductListingPage } from './ProductListingPage';

jest.mock('utils/useAPI');

const useAPIMock = useAPI as jest.Mock;

describe('<ProductListingPage />', () => {
  const mockProduct = {
    id: 123,
    title: 'Product title',
    artist: 'Artist name',
    isFree: true,
    image1: 'path/to/image1.jpg',
    accentColor: '#00b6b8',
    backgroundColor: '#F2F2F2',
    textColor: '#08070b',
  };

  const mockFetchProduct = jest.fn();

  const findByTestId = (wrapper: RenderResult, childTestId?: string) =>
    wrapper.getByTestId(getTestId('Products', childTestId));

  describe('@renders', () => {
    test('when loading', () => {
      useAPIMock.mockReturnValue([true, undefined, undefined, mockFetchProduct]);

      const wrapper = renderWithRouter(<ProductListingPage />);
      expect(wrapper.asFragment()).toMatchSnapshot();
    });

    test('with no products', () => {
      useAPIMock.mockReturnValue([false, [], undefined, mockFetchProduct]);

      const wrapper = renderWithRouter(<ProductListingPage />);
      expect(wrapper.asFragment()).toMatchSnapshot();
    });

    test('with a single product', () => {
      useAPIMock.mockReturnValue([
        false,
        [mockProduct],
        undefined,
        mockFetchProduct,
      ]);

      const wrapper = renderWithRouter(<ProductListingPage />);
      expect(findByTestId(wrapper, 'list')).toMatchSnapshot();
    });

    test('with multiple ordered products', () => {
      useAPIMock.mockReturnValue([
        false,
        [
          { ...mockProduct, id: 999, title: 'Product 2', feature_order: 2 },
          { ...mockProduct, id: 888, title: 'Product 3', feature_order: 3 },
          { ...mockProduct, id: 777, title: 'Product 1', feature_order: 1 },
        ],
        undefined,
        mockFetchProduct,
      ]);

      const wrapper = renderWithRouter(<ProductListingPage />);
      expect(findByTestId(wrapper, 'list')).toMatchSnapshot();
      expect(findByTestId(wrapper, 'carousel')).toMatchSnapshot('carousel');
    });

    test('with error', () => {
      useAPIMock.mockReturnValue([false, [], true, mockFetchProduct]);

      const wrapper = renderWithRouter(<ProductListingPage />);

      expect(wrapper.asFragment()).toMatchSnapshot();
    });
  });

  describe('@actions', () => {
    test('clicking try again button calls fetch products', () => {
      useAPIMock.mockReturnValue([false, [], true, mockFetchProduct]);

      const wrapper = renderWithRouter(<ProductListingPage />);

      expect(mockFetchProduct).not.toHaveBeenCalled();

      fireEvent.click(findByTestId(wrapper, 'tryAgain'));
      expect(mockFetchProduct).toHaveBeenCalledTimes(1);
    });
  });
});
