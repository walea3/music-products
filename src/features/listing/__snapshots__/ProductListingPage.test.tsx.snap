// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductListingPage /> @renders when loading 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Products
    </h1>
    <div>
      Loading products ...
    </div>
    <div>
      No products, yet.
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ProductListingPage /> @renders with a single product 1`] = `
<DocumentFragment>
  .c4 {
  display: block;
  width: 100%;
  height: auto;
  margin: 0;
  -webkit-transition: -webkit-transform 0.5s ease-in-out;
  -webkit-transition: transform 0.5s ease-in-out;
  transition: transform 0.5s ease-in-out;
}

.c2 {
  position: relative;
  display: inline-block;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-top: 5px solid;
  border-color: #00b6b8;
  overflow: hidden;
  width: 90%;
  max-width: 490px;
  color: #08070b;
}

.c2:hover .c3 {
  -webkit-transform: scale(1.2);
  -ms-transform: scale(1.2);
  transform: scale(1.2);
}

.c5 {
  position: relative;
  z-index: 1;
  padding: 20px;
  background-color: #F2F2F2;
  border-bottom-left-radius: 10px;
  border-bottom-right-radius: 10px;
}

.c6 {
  margin-top: 0;
}

.c7 {
  display: inline-block;
  right: 10px;
  padding: 5px 10px;
  border-radius: 5px;
  background-color: #00b6b8;
  font-weight: bold;
  font-size: 1.2rem;
}

.c0 {
  text-align: center;
  list-style: none;
  margin: 0 auto;
}

.c1 {
  display: inline-block;
  margin: 10px;
  vertical-align: top;
}

<div>
    <h1>
      Products
    </h1>
    <ul
      class="c0"
    >
      <li
        class="c1"
      >
        <a
          href="/product-123"
        >
          <article
            class="c2"
            data-testid="Product-123"
          >
            <img
              alt="Artist name"
              class="c3 c4"
              data-testid="Product-123-image"
              src="path/to/image1.jpg"
            />
            <div
              class="c5"
            >
              <h2
                class="c6"
                data-testid="Product-123-title"
              >
                Product title
              </h2>
              <div
                class="c7"
                data-testid="Product-123-price"
              >
                Free
              </div>
            </div>
          </article>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<ProductListingPage /> @renders with error 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Products
    </h1>
    There was problem getting you products. 
    <button
      data-testid="Products-tryAgain"
    >
      Try again.
    </button>
    <div>
      No products, yet.
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ProductListingPage /> @renders with multiple ordered products 1`] = `
<DocumentFragment>
  .c4 {
  display: block;
  width: 100%;
  height: auto;
  margin: 0;
  -webkit-transition: -webkit-transform 0.5s ease-in-out;
  -webkit-transition: transform 0.5s ease-in-out;
  transition: transform 0.5s ease-in-out;
}

.c2 {
  position: relative;
  display: inline-block;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-top: 5px solid;
  border-color: #00b6b8;
  overflow: hidden;
  width: 90%;
  max-width: 490px;
  color: #08070b;
}

.c2:hover .c3 {
  -webkit-transform: scale(1.2);
  -ms-transform: scale(1.2);
  transform: scale(1.2);
}

.c5 {
  position: relative;
  z-index: 1;
  padding: 20px;
  background-color: #F2F2F2;
  border-bottom-left-radius: 10px;
  border-bottom-right-radius: 10px;
}

.c6 {
  margin-top: 0;
}

.c7 {
  display: inline-block;
  right: 10px;
  padding: 5px 10px;
  border-radius: 5px;
  background-color: #00b6b8;
  font-weight: bold;
  font-size: 1.2rem;
}

.c0 {
  text-align: center;
  list-style: none;
  margin: 0 auto;
}

.c1 {
  display: inline-block;
  margin: 10px;
  vertical-align: top;
}

<div>
    <h1>
      Products
    </h1>
    <ul
      class="c0"
    >
      <li
        class="c1"
      >
        <a
          href="/product-777"
        >
          <article
            class="c2"
            data-testid="Product-777"
          >
            <img
              alt="Artist name"
              class="c3 c4"
              data-testid="Product-777-image"
              src="path/to/image1.jpg"
            />
            <div
              class="c5"
            >
              <h2
                class="c6"
                data-testid="Product-777-title"
              >
                Product 1
              </h2>
              <div
                class="c7"
                data-testid="Product-777-price"
              >
                Free
              </div>
            </div>
          </article>
        </a>
      </li>
      <li
        class="c1"
      >
        <a
          href="/product-999"
        >
          <article
            class="c2"
            data-testid="Product-999"
          >
            <img
              alt="Artist name"
              class="c3 c4"
              data-testid="Product-999-image"
              src="path/to/image1.jpg"
            />
            <div
              class="c5"
            >
              <h2
                class="c6"
                data-testid="Product-999-title"
              >
                Product 2
              </h2>
              <div
                class="c7"
                data-testid="Product-999-price"
              >
                Free
              </div>
            </div>
          </article>
        </a>
      </li>
      <li
        class="c1"
      >
        <a
          href="/product-888"
        >
          <article
            class="c2"
            data-testid="Product-888"
          >
            <img
              alt="Artist name"
              class="c3 c4"
              data-testid="Product-888-image"
              src="path/to/image1.jpg"
            />
            <div
              class="c5"
            >
              <h2
                class="c6"
                data-testid="Product-888-title"
              >
                Product 3
              </h2>
              <div
                class="c7"
                data-testid="Product-888-price"
              >
                Free
              </div>
            </div>
          </article>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<ProductListingPage /> @renders with no products 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Products
    </h1>
    <div>
      No products, yet.
    </div>
  </div>
</DocumentFragment>
`;
